### ISSUE TITLE
Ark+ finetune re-crashes fl-server in cross-site validation (prod 2026-07-15): #733 head-only eval broadcast not deployed; model re-orphaned (#727)

### ISSUE THREAD

**Issue by garciadias** [londonaicentre/FLIP#793]
## Summary

On **2026-07-15 (prod)** an Ark+ frozen-backbone fine-tune completed all 50 training rounds, then the
`fl-server` job process **crashed during the post-training cross-site validation** (`GlobalModelEval`)
phase — `child process exit with return code 1`. Because the crash landed before NVFLARE's `END_RUN`,
`PersistToS3AndCleanup` never ran: **no results were uploaded**, the model **orphaned at
`TRAINING_STARTED`**, and **net-1 stayed wedged** until an operator manually ran `POST /fl/stop`.

This is a fresh production recurrence of the crash/orphan pattern in **#727**, on the exact phase that
**#730 / #733** ("head-only broadcast for frozen-backbone cross-site validation") was written to fix —
**but #733 is not deployed to prod** (see [Root cause](#root-cause)). Filing to (a) get #733 rolled out,
(b) confirm it actually resolves this, and (c) add the observability + graceful-degradation that would
have made this a self-healing failed run instead of a manual recovery.

## Affected run

| | |
|---|---|
| Project | `529558aa-f1a0-465c-96c5-0dfd8b0ce8b4` ("Ark+ UK experiment") |
| Model | `98611b61-80a3-4596-bde0-40d112b6fffb` ("time metrics") |
| FL run / job | `ed7655b8-4240-4a42-88b4-e654105edc78` |
| Clients | Trust_1 (GSTT, fast link) + Trust_3 (BDMS/Thailand, slow link) |
| Backend | NVFLARE, net-1 |
| Deployed `fl-server` image | `ghcr.io/londonaicentre/flare-fl-server:7874061` (taskdef `fl-server-net-1:16`, registered 2026-07-09) |

## Timeline (from `/ecs/fl-server-net-1`, UTC)

```
17:51:36  Round 49 finished; final global model aggregated + persisted on server — training OK.
17:51:36  Workflow advances controller1 (ScatterAndGather) -> controller2 (GlobalModelEval).
17:51:37  GlobalModelEval loads SRV_server model, "Sending SRV_server model to all clients for validation".
17:51:39  validate task assigned to Trust_3 (task_id 78219920-...); model object streaming begins.
17:51:46  ServerEngine: Job ed7655b8 child process exit with return code 1   <-- CRASH
17:51:47  JobRunner: "Try to abort job ..." on clients (consequence of the crash, not the cause).
```

Model status after the crash (flip-api): stuck at **`TRAINING_STARTED`** (~24 h), status endpoint
returning `None` to the UI poller. `fl-api-net-1` shows only health polls after 17:51 — the
results-upload step was never reached.

## Trust_3 (client) is exonerated

Investigated on the Trust_3 node. The `fl-client` container **never crashed** (`ExitCode=0`,
`OOMKilled=false`, `Restarts=0`) and **never entered the `validate`/`cross_site_validate` branch**
(`trainer.py:519`) — it was blocked in `flare.receive()` (`trainer.py:490`) downloading the global
model when the server died. Its uploaded `error_log.txt` is a pure downstream symptom:

```
17:51:47 ArrayConsumer ERROR failed to download object with ref 84553e66-...: download aborted after 2.8s
17:51:47 Communicator ERROR Error getting task: failed to download from server.ed7655b8-...
17:51:48 InProcessClientAPI WARNING ask to stop job: reason: END_RUN received
```

No client-side traceback, no client OOM, no data-layer (XNAT catalog) errors. Trust_1 (fast link, no
in-flight download at the instant of death) uploaded no error log at all — both clients were innocent
bystanders.

## Root cause

The crash is **server-side, in the `GlobalModelEval` cross-site-validation broadcast**, and the
head-only eval broadcast that would prevent it **is not running in prod**:

1. **The deployed image predates the fix.** `fl-server-net-1:16` pins commit **`7874061`
   (2026-07-09 17:54)**. PR **#733** (head-only eval broadcast, closing #730) merged **`a1de4e15f`
   (2026-07-10 17:22)** — *after* the deployed commit. `git merge-base --is-ancestor a1de4e15f 7874061`
   → **not an ancestor**, i.e. **#733 is absent from the running image**. So the `validate` broadcast
   still ships the **full ~759 MiB global model** exactly as #730 described, instead of head-only.

2. **This is not the same clean OOM as #727, though.** #727's original occurrence used
   `CrossSiteModelEval` (submit_model all-to-all) and hit **96.8 % of the 8 GiB** task ceiling. This run
   already used **`GlobalModelEval`** (the #727 mitigation, deployed via #740) — `MemoryUtilized`
   peaked at only **~5.6 GiB / 8 GiB (~69 %)** through training and **never approached the ceiling**.
   Combined with **`rc=1` (a clean non-zero exit, not signal 137/-9)** and the crash being **fast
   (~9 s into the phase)**, the evidence points to an **uncaught Python exception during the
   full-model eval broadcast/stream**, not a hard task-level OOM. (A transient sub-minute allocation
   spike during full-model serialization can't be *excluded* — 1-min CloudWatch sampling has no
   datapoint inside the 9-s window — but there's no positive evidence for it.)

3. **The exact traceback is currently unrecoverable from the cloud side.** `/ecs/fl-server-net-1` has a
   single live stdout stream with no traceback (the job child's stack went only to the in-container
   `/app/<run>/log.txt`); **ECS exec is disabled** on the service + task; Fargate storage is ephemeral
   and the job dir is cleaned on abort.

## Proposed permanent solution

**Primary — deploy #733 to prod (removes the crash trigger).**
Roll the `fl-server` image forward to a commit that includes PR #733 (≥ `a1de4e15f`) so the `validate`
broadcast is head-only (only the `omni_heads` vars cross the wire; clients reconstruct the full model
from the frozen backbone they already hold). This deletes the 759 MiB eval stream that is the proximate
stressor. **Verify post-deploy** that (a) the fl-server logs a head-only trim on the `validate` task,
and (b) a full 50-round Ark+ finetune reaches `RESULTS_UPLOADED` with peak `MemoryUtilized` well below
the ~69 % seen here. This verification matters because the ~69 % / `rc=1` profile doesn't match a pure
OOM — we must confirm #733 actually closes *this* crash, not just the memory-pressure variant.

**Secondary — land #727 (makes any crash degrade gracefully).**
Even with #733, any future server-side death before `END_RUN` will silently re-wedge the net and
require manual `POST /fl/stop`. #727 (detect an orphaned/crashed run → set model `ERROR` → release the
net) is the defence-in-depth that turns this class of failure into a failed run instead of a stuck
queue. #733 and #727 are complementary: #733 removes this specific trigger; #727 catches all the others.

**Observability — capture the traceback next time.**
Enable ECS exec on `fl-server-net-1` (then read `/app/<run>/log.txt` before the abort cleanup) or stream
the job child's log to stdout/CloudWatch. Without this we cannot confirm the exact exception behind the
`rc=1`, which is why the primary fix needs the explicit post-deploy verification above. Note the
NVFLARE simulator (`job.py` SimEnv) likely will **not** reproduce this — it's prod/slow-link-specific
(staging on a single fast host never hit it).

## Recovery used for this incident

`POST /fl/stop/{model_id}` on `98611b61-…` → model terminal `STOPPED` → scheduler released net-1
(the flip-api internal ALB isn't externally reachable and ECS exec is off, so this was driven via a
one-off ECS `run-task` on the flip-api task definition calling `update_model_status(..., STOPPED)`).
Net-1 is now free.

## Related

- #727 — FL job crash before `END_RUN` orphans the model and wedges the net queue (OPEN; this incident is a fresh prod instance).
- #730 / #733 — head-only broadcast for frozen-backbone cross-site validation (the fix that is not yet deployed).
- #740 — Ark+ finetuning platform wiring (deployed the `GlobalModelEval` switch that lowered peak memory 97 % → 69 %).
